photom v2 roi selection for ablation demo

aaronalvarezcz opened the pull request:
This PR adds the feature that compromised the bulk of my work as an intern. It improves the PhotoM drawing mode capabilities, allowing the user to draw shapes/ROIs, fill them with patterns, and ablate over those patterns by interfacing with the hardware.

**Future Enhancement**
A good extension of this project is integrating it with on-the-fly segmentations and turning the segmentations into ROIs so they can be ablated. This will further automate and streamline the workflow for ablating with PhotoM.

aaronalvarezcz committed:
implement ROI selection for ablation via gui demo

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,263 @@
+from PyQt5.QtCore import Qt, QPoint, QRect, pyqtSignal
+from PyQt5.QtGui import QPainter, QPen, QPainterPath, QPolygon
+
+class Shape(QPolygon):
```
Let's make a new file inside `/copylot/assemblies/photom/utils/` (i.e pattern_tracing.py or similar). 
Make the necessary changes so that you import the Shape(QPolygon) class from that path.

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,263 @@
+        self.pattern_style = None
+        self.pattern_points = set()
+
+    def _pattern_bidirectional(self, vertical_spacing, horizontal_spacing):
```
can you add docstrings to the functions and the typing?
(i.e _pattern_bidirectional(self, int: vertical_spacing, int: horizontal_spacing)

edyoshikun replied:
this goes for all the functions that go into that utils/file.py

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,263 @@
+
+        # self.pattern_points.add((min_x, min_y))
+        # self.pattern_points.add((max_x, min_y))
+        # self.pattern_points.add((min_x, max_y))
```
remove unnecessary comments

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,263 @@
+        else:
+            self.show_or_hide_spacing_input(False)
+
+    def onAblateClick(self):
```
have this function output to a logger. `logger.debug(f'coords:{coords}')``

edyoshikun reviewed (commented):
This is a great demo of how we can draw and ablate some points. I wasn't able to check what the ablate method does. My suggestion is to get the copylot logger and use the `logging.logger.debug` or `logging.logger.info` to have this spit out information. For example, I would at least  `logger.info('Ablation Button Pressed')` and as `logger.debug(f'Ablation coordinates {coordinates of the shape to ablate})`)

I think this will be clear when you add the docstring, but it's not clear what does bidirectional means. 

Some things that would be nice to have is the logic to delete the specific shapes. So for example, have a method/wrapper that pops() or removes() specific shapes. You can add a button called Delete in the GUI that if the shape is selected from the dropdown and you click delete that it gets removed.

edyoshikun commented:
This one is minor but if the spacing is too big the behaviour should be that it at least has 1 point in the ROI and also throw some `warn` and/or `logger.warning`

aaronalvarezcz committed:
move Shape class to utils/pattern_tracing

aaronalvarezcz committed:
add docstrings and logging statements

aaronalvarezcz committed:
add roi selection coloring in the GUI

aaronalvarezcz committed:
fix roi selection coloring bug

aaronalvarezcz committed:
add shape deletion logic and button

aaronalvarezcz committed:
remove unnecessary imports and update docstrings

aaronalvarezcz committed:
add default behavior for overly large spacing configuration

edyoshikun commented on copylot/assemblies/photom/utils/pattern_tracing.py:
```diff
@@ -0,0 +1,59 @@
+from PyQt5.QtGui import QPolygon
+import logging 
+
+class Shape(QPolygon):
```
let's rename this to something more meaningful as well. trace_shape or similar.

aaronalvarezcz replied:
you mean renaming the shape class -> trace_shape?

edyoshikun replied:
@aaronalvarezcz address this. Let's not name it shape because that is too common of a name and is used in a bunch of other packages. Name it  consistent to what you have named the other functions. (i.e Shape_tracer, shape_drawer,etc)

edyoshikun replied:
still pending@aaronalvarezcz

edyoshikun committed:
adding changes to be able to run demo_mode

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,314 @@
+        )
+
+        # roi dropdown box
+        self.roi_dropdown= QComboBox(self)
```
these need to be added to the control window in `gui.windows.py`

edyoshikun replied:
all the new buttons and functionality should show up in the photom GUI when you enable 'drawing mode'.

edyoshikun replied:
and should dissapear otherwise.

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,314 @@
+                point = event.pos()
+                self.curr_shape_points.append(point)
+
+                self.curr_shape_points.append(self.curr_shape_points[0]) # connecting the final points in case not connected
```
it might be that this needs to be abstracted a bit so it's easier to read. You will need to add this to the `eventFilter`in windows.py

edyoshikun commented on copylot/assemblies/photom/demo/demo_roi.py:
```diff
@@ -0,0 +1,314 @@
+        self.curr_shape_id = 0
+        self.selected_shape_id = None
+
+    def mousePressEvent(self, event: QMouseEvent) -> None:
```
these mouseEvents should follow comment below.

aaronalvarezcz committed:
Integrate ROI selection to Photom

Added:
- drawing mode group in control window, includes roi selection and pattern dropdown, pattern spacing boxes, and deletion of shapes
- scrollable widget behavior for laser control window
- attributes to LaserMarkerWindow class to contain shapes

Changed:
- how displaying shapes are handled. shape/drawings are displayed by calling self.update() now
- shapesUpdate signal to notify of new shapes added
- updated mouse event handling to track and store Shape objects

edyoshikun commented on copylot/assemblies/photom/gui/windows.py:
```diff
@@ -136,7 +139,7 @@ def initialize_UI(self):
             self.photom_window_pos[1],
             self.photom_window_size_x,
-            self.photom_window_size_x,
+            1000,
```
let's not hardcode this variable. If needed, change the `photom_gui.py` file that executes this PhotomApp window.

edyoshikun commented on copylot/assemblies/photom/gui/windows.py:
```diff
@@ -215,10 +218,83 @@ def initialize_UI(self):
+        self.spacing_boxes.addWidget(self.horizontal_spacing_label)
+        self.horizontal_spacing_input = QLineEdit(self)
+        self.spacing_boxes.addWidget(self.horizontal_spacing_input)
+
```
```suggestion
```

edyoshikun replied:
you can format this with black to avoid these extra spaces and inconsistent use of Quotes ("") vs ('')

edyoshikun commented on copylot/assemblies/photom/gui/windows.py:
```diff
@@ -582,7 +737,7 @@ def initMarker(self):
         # self.marker.setBrush(QColor(255, 0, 0))
         # Load the PNG image
-        pixmap = QPixmap(r'./copylot/assemblies/photom/utils/images/hit_marker_red.png')
+        pixmap = QPixmap(r'/Users/aaron.alvarez/Desktop/biohub/coPylot/copylot/assemblies/photom/utils/images/hit_marker_red.png')
```
Let's make sure this is a relative import to the current file.

edyoshikun commented on copylot/assemblies/photom/gui/windows.py:
```diff
@@ -879,6 +1027,27 @@ def _move_marker_and_update_sliders(self):
             ].mirror_y_slider.setValue(new_coords[1][0])
+    
+    def _roi_tracing(self):
+        if self.selected_shape_id != None:
```
is not 
```suggestion
        if self.selected_shape_id is not None:
```

edyoshikun commented on copylot/assemblies/photom/gui/windows.py:
```diff
@@ -879,6 +1027,27 @@ def _move_marker_and_update_sliders(self):
+            shape = self.shapes[self.selected_shape_id]
+            new_pattern_points = []
+            if shape.pattern_points:
+                for position in shape.pattern_points:
```
currently these will be fed without a delay. Let's add a pattern_delay variable that we can set to pause between positions.

edyoshikun replied:
the default can be zero.

aaronalvarezcz committed:
remove hardcoding control window height

aaronalvarezcz committed:
use relative path for hit marker png

aaronalvarezcz committed:
fix comparison check if there is a selected shape id

aaronalvarezcz committed:
add pattern_delay time b/w positions for mirror movement

aaronalvarezcz committed:
format with black

edyoshikun committed:
-adding changes to make these relative imports.
-relative import for the crosshair

edyoshikun commented:
When I draw the different ROIs, and click apply. It only moves to position [0] and it freezes. It doesn't print the next positions. Do you have an idea on why that happens?

edyoshikun committed:
extending the range of mockmirrodevice

aaronalvarezcz committed:
refactor Shape class to ShapeTrace

aaronalvarezcz committed:
make config_file a path object

aaronalvarezcz committed:
add n_points option for pattern generation on a shape

aaronalvarezcz committed:
formatted with black

aaronalvarezcz committed:
add spiral pattern generation code for ROIs

aaronalvarezcz committed:
add gap and points per cycle attributes to ShapeTrace

aaronalvarezcz committed:
remove spacing option for spiral pattern in gui

aaronalvarezcz committed:
update _pattern_spiral to generate n equidistant points in a spiral

aaronalvarezcz committed:
update bidirectional pattern names for better spacing

aaronalvarezcz committed:
implement bfs-style bidirectional pattern application

aaronalvarezcz committed:
change pattern_points from set to list for ordered points

edyoshikun committed:
add greyscale pattern tracing, change pattern_points from set to list

edyoshikun committed:
add time delay parameter in GUI

aaronalvarezcz committed:
remove unnecessary code and comments

aaronalvarezcz committed:
fix pattern drawing bug

aaronalvarezcz committed:
update breadth first search to add points clockwise to queue

aaronalvarezcz committed:
fix bidirectional pattern bug that allowed for repeated (x, y) coordinates

edyoshikun reviewed (approved):
LGTM! Thanks for these new features @aaronalvarezcz. This looks great, so I am merging it.
